### PR TITLE
CosmosDB- Add JsonReader, JsonWriter, and JsonNavigator Interfaces

### PR DESCRIPTION
**Pull request by Maya-Painter**
# Description

This introduces JSON reader, writer, and navigator interfaces for CosmosDB.  

This is part of the effort to support the CosmosDB binary format on the client side. Upcoming PRs will include implementations for text and binary encodings.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by alzimmermsft** [comment]
Hey @Maya-Painter, this PR appears very similar to functionality available in `azure-json`s APIs ([JsonReader](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-json/src/main/java/com/azure/json/JsonReader.java), [JsonWriter](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-json/src/main/java/com/azure/json/JsonWriter.java)), albeit with some more convenience methods for more Java types. We're also working on adding JSON pointer functionality into `azure-json` which it seems this PR is also doing as well, alongside adding stream-style JSON APIs.

Is it possible for Cosmos to leverage `azure-json` instead of needing to maintain these APIs? Is there a gap we can work to close in `azure-json` to make this adoption possible?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+     */
+    Iterable<ObjectProperty> getObjectProperties(IJsonNavigatorNode objectNode);
+
+    /**
+     * Creates an {@link IJsonReader} that is able to read the supplied {@link IJsonNavigatorNode}.
+     *
+     * @param jsonNavigatorNode The node to create a reader from.
+     * @return The {@link IJsonReader} that is able to read the supplied {@link IJsonNavigatorNode}.
+     */
+    IJsonReader createReader(IJsonNavigatorNode jsonNavigatorNode);
```
Comment (sboshra) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java:174]: >jsonNavigatorNode [](http://example.com/codeflow?start=48&length=17)

node

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+    /**
+     * Writes a {@link IJsonNavigatorNode} to a {@link IJsonWriter}.
+     *
+     * @param jsonNavigatorNode The {@link IJsonNavigatorNode} to write.
+     * @param jsonWriter        The {@link IJsonWriter} to write to.
+     */
+    void writeNode(IJsonNavigatorNode jsonNavigatorNode, IJsonWriter jsonWriter);
```
Comment (sboshra) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java:182]: >jsonNavigatorNode [](http://example.com/codeflow?start=38&length=17)

node

Comment (sboshra) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java:182]: >jsonWriter [](http://example.com/codeflow?start=69&length=10)

writer

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * Corresponds to a UUID type in Java for the extended types.
+     */
+    UUID,
+
+    /**
+     * Unknown JsonNodeType.
+     */
+    UNKNOWN,
+}
```
Comment (sboshra) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java:93]: We should not add this

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonSerializationFormat.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public static JsonSerializationFormat fromByte(byte value) {
+        for (JsonSerializationFormat enumValue : JsonSerializationFormat.values()) {
+            if (enumValue.byteValue == value) {
+                return enumValue;
+            }
+        }
+        throw new IllegalArgumentException("No JsonSerializationFormat with byte value " + value);
+    }
```
Comment (sboshra) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonSerializationFormat.java:45]: missing '.' at the end of the sentence.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+package com.azure.cosmos.implementation.json;
+
+/**
+ * The enumeration of JSON node types.
+ */
+enum JsonNodeType {
+    /**
+     * Corresponds to the 'null' value in JSON.
+     */
+    NULL,
```
Comment (sboshra) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java:12]: >NULL [](http://example.com/codeflow?start=4&length=4)

Is all caps the naming convention here?

Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java:12]: It's the Java convention and seems to be followed in some (but not all) of the SDK enum types.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java (hunk 3 above)
Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java:88]: removed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java (hunk 5 above)
Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java:12]: @kushagraThapar any preference here?

**Comment by kushagraThapar** [comment]
@Maya-Painter / @adityasa - I had a conversation with @alzimmermsft and it makes sense to proceed with the review of this PR as it is without depending on azure-core/azure-json library interfaces / implementation for now. 
Once azure-core / azure-json ships these interfaces, we will revisit our design and see how we can leverage their interfaces to avoid duplicate code in the future. 

@alzimmermsft - please feel free to clarify more if I missed anything, thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Corresponds to the byte type in Java for the extended types.
+     */
+    INT8,
+
+    /**
+     * Corresponds to the short type in Java for the extended types.
+     */
+    INT16,
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java:57]: I curious should these all be different JsonNodeType or categorize them into Number

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/ObjectProperty.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+class ObjectProperty {
+
+    /**
+     * Initializes a new instance of the ObjectProperty struct.
+     *
+     * @param nameNode The IJsonNavigatorNode to the node that holds the object property name.
+     * @param valueNode The IJsonNavigatorNode to the node that holds the object property value.
+     */
+    public ObjectProperty(IJsonNavigatorNode nameNode, IJsonNavigatorNode valueNode) {
+        this.nameNode = nameNode;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/ObjectProperty.java:17]: add null check?

Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/ObjectProperty.java:26]: added check

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java (hunk 6 above)
Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java:57]: The intention with this model is to create a 1:1 mapping between node types and return value types.  Especially in the case of the Navigator classes (ex: https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Json/JsonNavigator.JsonBinaryNavigator.cs#L526) it should simplify implementing write functions.  I would argue to leave it as is for now, and then review if it makes sense to combine these into one enum value when we revaluate aligning with the azure-core/azure-json library interfaces.

Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonNodeType.java:57]: The intention with this model is to create a 1:1 mapping between JsonNodeTypes and Java types. Especially in the case of the Navigator classes (ex:  https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Json/JsonNavigator.JsonBinaryNavigator.cs#L526) it should simplify implementing some functions. I would argue to leave it as is for now, and then review if it makes sense to combine these into one enum value when we revaluate aligning with the azure-core/azure-json library interfaces.

**Review by xinlian12 — APPROVED** [review]
LGTM,thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    ByteBuffer getBinaryValue(IJsonNavigatorNode binaryNode);
+
+    /**
+     * Tries to get the buffered binary value from a node.
+     *
+     * @param binaryNode          The node to get the buffered binary from.
+     * @param bufferedBinaryValue The buffered binary value if possible.
+     * @return {@code true} if the JsonNavigator successfully got the buffered binary value; {@code false} if the JsonNavigator failed to get the buffered binary value.
+     */
+    boolean tryGetBufferedBinaryValue(IJsonNavigatorNode binaryNode, ByteBuffer bufferedBinaryValue);
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java:115]: Are you trying to mimic an out byte[] buffer here? If so, AtomicReference<ByteBuffer> probably works better?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonTokenType.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+/**
+ * Enumeration of JSON token types.
+ */
+enum JsonTokenType {
+
+    /**
+     * Reserved for no other value.
+     */
+    NOTSTARTED,
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonTokenType.java:13]: NIT - I would prefer either using CamelCase as in .Net or if you use all-caps separate words with a _

In general I think sticking to what you have in .Net is fine. This is an implementation package API - will never be in public surface area - so, we have some liberty on how to use naming conventions.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/ObjectProperty.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * Initializes a new instance of the ObjectProperty class.
+     *
+     * @param nameNode The IJsonNavigatorNode to the node that holds the object property name.
+     * @param valueNode The IJsonNavigatorNode to the node that holds the object property value.
+     */
+    public ObjectProperty(IJsonNavigatorNode nameNode, IJsonNavigatorNode valueNode) {
+
+        if (nameNode == null) {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/ObjectProperty.java:18]: Please use checkNotNull precondition function

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM except for few minor comments

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonTokenType.java (hunk 9 above)
Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/JsonTokenType.java]: Updated all enums to be PascalCase

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/ObjectProperty.java (hunk 10 above)
Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/ObjectProperty.java]: Updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java (hunk 8 above)
Comment (Maya-Painter) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/json/IJsonNavigator.java]: Updated all references of ByteBuffer to AtomicReference<byte[]>

**Pull request merged**